src/utils/style-utils.js

```diff
@@ -1,137 +1,72 @@
-    if (prevKeysList.length !== nextKeysList.length) {
-      return true;
-    }
-    // `nextStyle` and `prevStyle` should not have the same set of props.
-    if (nextKeysList.some(
-      key => prevStyle.get(key) !== nextStyle.get(key)
-      // But the value of one of those props is different.
-    )) {
-      return true;
-    }
-    return false;
+  if (typeof style === 'string') {
+    return null;
   }
-
-  if (!prevStyle || propsOtherThanLayersOrSourcesDiffer()) {
-    map.setStyle(nextStyle.toJS());
-    return;
+  if (style.toJS) {
+    style = style.toJS();
   }
 
-  const {sourcesDiff, layersDiff} = diffStyles(prevStyle, nextStyle);
-  checkForEqualLayerSourceChanges(sourcesDiff.exit, nextStyle.get('layers'));
-  applySourceLayerChanges(map, nextStyle, sourcesDiff, layersDiff);
-}
-
-/* eslint-enable max-statements, complexity */
-
-// Update a source in the map style
-function updateStyleSource(map, update) {
-  const newSource = update.source.toJS();
-  if (newSource.type === 'geojson') {
-    const oldSource = map.getSource(update.id);
-    if (oldSource.type === 'geojson') {
-      // update data if no other GeoJSONSource options were changed
-      const oldOpts = oldSource.workerOptions;
-      // GeoJSONSource class scales user options before assigning to workerOptions
-      // https://github.com/mapbox/mapbox-gl-js/blob/master/src/source/geojson_source.js
-      const scale = oldOpts.geojsonVtOptions.extent / 512;
-
-      if (
-        (newSource.maxzoom === undefined ||
-          newSource.maxzoom === oldOpts.geojsonVtOptions.maxZoom) &&
-        (newSource.buffer === undefined ||
-          newSource.buffer === oldOpts.geojsonVtOptions.buffer / scale) &&
-        (newSource.tolerance === undefined ||
-          newSource.tolerance === oldOpts.geojsonVtOptions.tolerance / scale) &&
-        (newSource.cluster === undefined ||
-          newSource.cluster === oldOpts.cluster) &&
-        (newSource.clusterRadius === undefined ||
-          newSource.clusterRadius === oldOpts.superclusterOptions.radius / scale) &&
-        (newSource.clusterMaxZoom === undefined ||
-          newSource.clusterMaxZoom === oldOpts.superclusterOptions.maxZoom)
-      ) {
-        oldSource.setData(newSource.data);
-        return;
-      }
-    }
+  if (Array.isArray(style.layers)) {
+    return style.layers.filter(l => l.interactive).map(l => l.id);
   }
 
-  map.removeSource(update.id);
-  map.addSource(update.id, newSource);
+  return null;
 }
 
-function applySourceLayerChanges(map, nextStyle, sourcesDiff, layersDiff) {
-// TODO: It's rather difficult to determine style diffing in the presence
-  // of refs. For now, if any style update has a ref, fallback to no diffing.
-  // We can come back to this case if there's a solid usecase.
-  if (layersDiff.updates.some(node => node.layer.get('ref'))) {
-    map.setStyle(nextStyle.toJS());
-    return;
+export function normalizeStyle(style) {
```

Comment:
It would be nice with a comment header that explains the various things we need to do as part of "normalization".